[Docs] generate include files to ensure referenced Github branch matches docs

djptek opened the pull request:
- Have you signed the [contributor license agreement](https://www.elastic.co/contributor-agreement)? Y
- Have you followed the [contributor guidelines](https://github.com/elastic/ecs/blob/master/CONTRIBUTING.md)? Y
- For proposing substantial changes or additions to the schema, have you reviewed the [RFC process](https://github.com/elastic/ecs/blob/master/rfcs/README.md)? n/a
- If submitting code/script changes, have you verified all tests pass locally using `make test`? Y
- If submitting schema/fields updates, have you generated new artifacts by running `make` and committed those changes? n/a
- Is your pull request against master? Unless there is a good reason otherwise, we prefer pull requests against master and will backport as needed. Y
- Have you added an entry to the [CHANGELOG.next.md](https://github.com/elastic/ecs/blob/master/CHANGELOG.next.md)? Y

This addresses comment https://github.com/elastic/ecs/issues/1313#issuecomment-807630113 which identified that versions quoted in asciidocs are out of sync with referenced github trees, for example the `ECS Github repo` link on 
the page 

https://www.elastic.co/guide/en/ecs/1.8/ecs-artifacts.html

references

https://github.com/elastic/ecs/blob/master/generated#artifacts-generated-from-ecs

Proposed solution: add logic to `generate.py` to create intermediate includes:

```
generated/ecs_github_repo_link.asciidoc
generated/ecs_version.asciidoc
```
to ensure `ecs_version` is aligned with `ecs_github_repo_link`

tested OK with:

```
version   | expected result                | actual result
----------+--------------------------------+-------------------------------
2.0.0-dev | text 2.0.0-dev/links to master | links to master
1.9.0     | text/links to 1.9.0            | text/ (valid) links to 1.9.0 
9.9.9     | text/links to 9.9.9            | text/ (invalid) links to 9.9.9 
```
Note: this is a different route to how `generators/asciidoc_fields.py` adds the version, I think that's OK as this targets a different context, handwritten `asciidoc` files rather than python generated `asciidoc`

Questions to reviewer: are the generated asciidoc files in the right place?

djptek commented:
Tests passing for https://github.com/elastic/ecs/pull/1338/commits/29e03d6a5b2d139bf950826398d967f02bd1daf6  ONLY because I checked in the two output includes to get visibility of the build in Jenkins. removing those now

djptek commented:
To facilitate the review, I've set 

`echo "1.10.0\n" > version`

which results in a link to: 

https://github.com/elastic/ecs/tree/1.10/generated#artifacts-generated-from-ecs

that is currently a `404`, however if you check e.g. 

https://github.com/elastic/ecs/tree/1.9/generated#artifacts-generated-from-ecs

it is apparent the link will be valid as and when the `1.10` docs are published

If this PR is accepted I can change that back to `2.0.0-dev` which will yield a link to 

https://github.com/elastic/ecs/tree/master/generated#artifacts-generated-from-ecs

which is likewise valid

Note: the current `elasticsearch-ci/docs` process appears not to run `make` before `make docs`, so I added this 
step and ensured that 

```
docs/ecs_github_repo_link.asciidoc
docs/ecs_version.asciidoc
```

were explicitly added to the commit, which is currently a manual step, which would need to be added to the release process &/or automated.

ebeahan reviewed (commented):
Sorry for taking a bit to review this one.

I looked into some of the available docs attributes, and I stumbled upon `{source_branch}`. I only tested a bit, but I believe it'll function the same as the changes proposed here?

```asciidoc
:ecs_github_repo_link: https://github.com/elastic/ecs/blob/{source_branch}
```

djptek commented on Makefile:
```diff
@@ -44,7 +44,7 @@ codegen: gocodegen
 # Build the asciidoc book.
 .PHONY: docs
-docs:
+docs: generator
```
Thanks @ebeahan `{source_branch}` looks promising, I'll see if I can make that work with less intervention likewise adding a new target

djptek replied:
@ebeahan Moving back to draft while I work on this - the fields `{source_branch}` and also `{version}` are already available so I should use those rather than building (potentially desynchronized) alternatives, however, their underlying logic is a little unclear to me: for example if I set the file 

`ecs/version` to `1.10.0`

and then generate the docs

then 

`version` evaluates to `1.8`

and

`source_branch` evaluates to `master` 

Which is not what I would have (initially) expected

I'm looking into the upstream bash scripts to see whether they need fixing or I have simply misinterpreted them  + evaluate whether any other code/docs relies on their current behavior

djptek replied:
Please see summary in https://github.com/elastic/ecs/pull/1338#issuecomment-824305976

djptek commented:
`gendocs` is a great idea, so I've applied this, yay!

`source_branch` lacks the logic to deal with 2.0.0-dev => master, so I kept that include

Also, I looked at moving `ecs_github_repo_link` to `attributes.asciidoc` however, again asciidoc doesn't provide for that logic so I kept the python logic and wrote an include. I also moved `tree` to `blob`

The conflict with `ecs_version` arose because this was read from the file `./version` and is therefore distinct to `ecs_version` as defined in `attributes.asciidoc`. Again, there is logic here, on one side adding `-exp` for the generator also removing `-dev` for the docs, so again I kept the python logic and renamed references to the generated `ecs_version` to `ecs_generated_version` for clarity

@ebeahan as this is following the Feature Freeze, I'm getting a conflict with `CHANGELOG.next.md` and could appreciate guidance there

ebeahan commented on CHANGELOG.next.md:
```diff
@@ -30,6 +30,8 @@ Thanks, you're awesome :-) -->
 
 ### Tooling and Artifact Changes
 
+* Generate include files to ensure referenced Github branch matches docs #1338
```
> @ebeahan as this is following the Feature Freeze, I'm getting a conflict with CHANGELOG.next.md and could appreciate guidance there

Yeah, that can happen if the `CHANGELOG.next` file is updated while you're working on your branch here.

Looking at the changelog entry here, we actually want to place it under one of the section headers underneath `Tooling and Artficact Changes`. Perhaps `Improvements`? Moving the entry might also fix the conflict.

djptek replied:
OK, I moved it back there :+1:

djptek commented on scripts/generator.py:
```diff
@@ -102,5 +104,16 @@ def read_version(ref=None):
+    with open(os.path.join(directory, 'ecs_version.asciidoc'), 'w') as version_include:
+        version_include.write(':ecs_version: ' + version)
+    with open(os.path.join(directory, 'ecs_github_repo_link.asciidoc'), 'w') as github_include:
+        lhs = ':ecs_github_repo_link: https://github.com/elastic/ecs/tree/'
```
went with blob, as that seems to be what the code uses elsewhere: Github resolves this to tree later

djptek commented:
I backed out the changes to `Makefile` as I'm not happy with how that is working, currently requires:

```
$ make
$ make docs
$ make check
$ git status 
...
experimental/generated/csv/fields.csv: needs update
...
generated/csv/fields.csv: needs update
...
$ git add -A
$ git commit -m "XXX" 
$ git push
```

to pass `elasticsearch-ci/docs`

Note: `make ; make docs` creates the intended docs including e.g. 

`generated/csv/fields.csv: needs update`

*but* does not generate  

`experimental/generated/csv/fields.csv: needs update`

So, I'm opening another ticket to look into that TBC

ebeahan commented:
As I understand it, there are two items this PR is tackling:

1. Map the `version` file's version to the GitHub branch associated with that version.
2. Automatically drop in the version in `docs/index.asciidoc` based on the contents of the `version` file.

I'm going to cover both below, but let me know if I'm misunderstanding the intent here.

## Linking to GitHub branches

I still believe the `{source_branch}` attribute could solve the first problem. `{source_branch}` is based on the git branch the documentations is being built from for ECS, and it should always link accurately back to the appropriate git branch.

I tested defining:

`:ecs_github_repo_link: https://github.com/elastic/ecs/blob/{source_branch}`

and then using that attribute to create a URL to the README in the various branches (`master`, `1.x`, `1.10`):

`This is the link to the GitHub repo: {ecs_github_repo_link}/README.md`

As I built the docs in the different branches, the value filled in by `{source_branch}` pointed to the expected location in GitHub:

`master`

![Screen Shot 2021-04-22 at 4 31 32 PM](https://user-images.githubusercontent.com/7226265/115788406-2f0a8e00-a389-11eb-82aa-65b74508c520.png)

`1.x`

![Screen Shot 2021-04-22 at 4 35 47 PM](https://user-images.githubusercontent.com/7226265/115788471-421d5e00-a389-11eb-9abc-eb5376f0178e.png)

`1.10`

![Screen Shot 2021-04-22 at 4 36 25 PM](https://user-images.githubusercontent.com/7226265/115788514-4f3a4d00-a389-11eb-9718-7d824b58a092.png)

Is this the type of result we're looking for?

## `doc/index.asciidoc` version

It would be a plus not to have to edit this manually every time we bump the versions. 😄 

There is another spot where this is already been done in the `scripts/generators/asciidoc_fields.py` generator, where `ecs_version` is plugged into the Jinja2 template:

https://github.com/elastic/ecs/blob/master/scripts/templates/fields.j2#L5

Reusing that approach on `index.asciidoc` would allow a single location to update the version in the `version` file.

djptek commented:
@ebeahan 

>>I tested defining:

>>:ecs_github_repo_link: https://github.com/elastic/ecs/blob/{source_branch}

>>and then using that attribute to create a URL to the README in the various branches (master, 1.x, 1.10):

Did you push to GitHub & go through the build process?

I agree 100% that source branch works in local, however, when I pushed to GitHub, for example, with source at 1.10.0 the build process actually built refs to 1.8.0 (!) as explained above

This is why I opened #1378 in order to better understand

An alternative might simply be to solve this through convention by adding a bullet point to mandate `make check` in the PR template as a precursor to generating new artifacts.

I can roll back the includes and re-implement with `source_branch` or potentially other values from `attributes.asiidoc` , however first I'd ratification of the logic relating the github version to the displayed version which I understand matches the actual `./version` file:

```
./version | generated version | github version | stack version | master/asciidoc:ecs_version
----------+-------------------+----------------+---------------+----------------------------
2.0.0-dev | 2.0.0-dev         | master         | master        | 1.9
2.0.0-dev | 2.0.0-dev+exp     | master         | master        | 1.9
1.10.0    | 1.10.0            | 1.10.0         | tbc           | tbc
1.9.0     | 1.9.0             | 1.9.0          | 7.12          | 1.8
```

djptek commented:
Closing, superseded by [Docs] Fix GitHub repo link using source_branch + templates #1381
